Rework flight check

flaviojs opened the pull request:
Reference: #2944

**Changes in this PR**
- add warning and error icons (by MZ)
- reorganize ship warnings
- display a packed ship warning icon with a multi-warning tooltip per ship in the list of ships of the ShopPanel and the PlayerInfoPanel.
- display all ship warnings as icons with tooltips in the ship details of the ShipPanel and the ShipInfoPanel.
- add new class ShipWarnings.
- add new class TextTooltip
- move flight checks from the outfitter to take off
- show 1 conversation per ship warning and allow 4 actions:
  - go to the outfitter: available if you can access the outfitter and the ship is in the system
  - abort the take off: always available
  - park the ship: available if it's not be the flagship
  - dismiss the warning: available if it's not a serious warning

flaviojs commented:
Screenshot of the flight check line and tooltip (with engines removed):
![flight_check](https://user-images.githubusercontent.com/1009600/30330046-e314e8d2-97cb-11e7-96df-da33a5331d0e.png)

tehhowch commented on source/ShipInfoDisplay.cpp:
```diff
@@ -303,6 +318,27 @@ void ShipInfoDisplay::UpdateAttributes(const Ship &ship, const Depreciation &dep
 	attributesHeight += 30;
+	
+	// Flight check warnings
+	if(ship.IsYours())
```
Is this check needed? Surely one would want to know if there are flight warnings for any ship they can see the information of?

flaviojs replied:
The check makes ships for sale in the shipyard (or the map's shipyard) not show that particular line.

tehhowch replied:
Yes, I'm aware. If a stock ship being sold is not ready to fly, that should be communicated to the player. This is beneficial for plugin-makers, as a content sanity check, or as a gameplay mechanic (e.g. selling discount ships which are missing parts).

flaviojs replied:
Good point, will remove the check.

tehhowch commented on source/ShipInfoDisplay.cpp:
```diff
@@ -303,6 +318,27 @@ void ShipInfoDisplay::UpdateAttributes(const Ship &ship, const Depreciation &dep
+	// Flight check warnings
+	if(ship.IsYours())
+	{
+		vector<string> warnings = ship.FlightCheck();
```
Why make FlightCheck() a Ship class member, rather than a ShipInfoDisplay member? The Ship class is pretty large already, and the game only uses FlightCheck() for informational purposes - the rest of Ship class members are used to instantiate ships or help them make decisions in flight (e.g. used by the Ship, AI, and Engine classes).

flaviojs replied:
Because it will be used in other places (this is a work in progress) and it just makes sense.
The Ship class has methods that provide information about the ship. I consider this function to be the same kind of function and the consequences of this information is left to outside code.

tehhowch replied:
I suppose if you are planning to reference this data from PlayerInfoPanel (which ships have warnings), PlanetPanel/PlayerInfo (can we take off?) or OutfitterPanel/ShopPanel (can we leave the outfitter?), then it being a Ship class member makes sense.

flaviojs commented:
Just to make it clear, the WIP in the title means this is a work in progress and is not ready to be merged. The reason it's already a PR is so I can get feedback.

I want this PR to address all the flight check stuff I mentioned in #2944.

Bladewood commented:
I'd suggest indenting the individual tasks or bulleting them (somehow).

endless-sky commented:
It might be better to have a warning icon appear above the ship's icon in the outfitter, rather than putting this at the bottom of the info display, for two reasons:

1. If it's at the bottom of the info display, it will overflow the height of the ship info panel.
2. If you have a lot of ships, an icon would make it easier to see which ship has issues without clicking through all of them.

flaviojs commented:
A warning icon (or another visual cure) above the ship icon is also planned.

I checked ShipInfoPanel before committing this and it looked fine. Originally I wanted to show the problems too but it didn't fit ShipInfoPanel so I moved them to the tooltip.

Screenshot of left edges in ShipInfoPanel:
![flight_check_shipinfopanel](https://user-images.githubusercontent.com/1009600/30372697-8f320190-9876-11e7-9cd4-790bc3d61f87.png)

flaviojs commented:
@Bladewood
WrappedText doesn't support indenting paragraphs so I'll do simple bulleting.
![flight_check_2](https://user-images.githubusercontent.com/1009600/30373832-966a9f86-987a-11e7-9231-2bea8621fdd5.png)

Amazinite commented:
I'd definitely prefer this over the current system if finalized.

tehhowch commented:
As of 1e1a6fc, the conversations are still shown when leaving the outfitter: https://github.com/endless-sky/endless-sky/pull/3040/files#diff-b6734b3a6bc3d5b29d3c6a3cd3af1a75R622

I think the intent is to move those to when trying to leave the planet, which will allow the player to park empty hulls, or decide to sell a hull after selling engines/critical systems (with the original intent to upgrade) without needing to rebuy all missing critical systems.

flaviojs commented:
Indeed, I intend to move those conversations to when trying to leave the planet.
Before that I'll need to thing about how multiple ship problems and multiple problematic ships should be handled by the conversation.
And at an even more basic level, I'll need to learn about the capabilities of conversations. xD

Found that the flight check line overflows in ShipInfoPanel when there are fighter/drone bays so I'll need to revisit it.

flaviojs commented:
For the list with your ships in the shop, I think this is enough:
![flight_check_your_ships](https://user-images.githubusercontent.com/1009600/30514962-72051410-9b17-11e7-8d19-a3cef726fd25.png)

For the flight check details, I tried a bunch of stuff to see what would work on the various places and I liked best putting icons bellow the ship image and more details in tooltips.

Here are the prototypes with text instead of icons:
![flight_check_3](https://user-images.githubusercontent.com/1009600/30514975-c6e60d2c-9b17-11e7-867a-788d787124c1.png) ![flight_check_shipinfopanel_2](https://user-images.githubusercontent.com/1009600/30514978-ce79ad5a-9b17-11e7-8c84-0872c70763eb.png)

The text maps to these flight check warnings:
* E - "no energy"
* T - "no thrusters"
* TE - "no thruster energy"
* S - "no steering"
* SE - "no steering energy"
* SP - "solar power warning"
* O - "overheating"

If this placement is fine, can someone make an icon for each flight check warning? xD

flaviojs commented:
@endless-sky
In order to use text tooltips in other places I'm moving relevant code from ItemInfoDisplay to a new class TextTooltip. What should I put in the copyright, keep MZ there and update to 2017?

Bladewood commented on data/tooltips.txt:
```diff
@@ -289,6 +289,37 @@ tip "max:"
+# Ship warnings:
+tip "flight check: no energy"
+	`Insufficient energy is being generated or stored.`
+	`Install energy sources or batteries`
```
missing period

flaviojs replied:
Fixed.

flaviojs commented:
I updated the project files but they are untested (I don't use either).

flaviojs commented:
Forgot to mention this in the log, ShipWarnings also detects when no energy is being produced (normally and far away from the sun).

endless-sky commented:
There are so many changes involved here that it's a bit hard to review them and decide if they all make sense or not. It might make sense to figure out the design details first, then implement the changes as simply as possible.

Note that commit ba06025 made it possible for thrusters and steering to operate, but at reduced effectiveness, if there's not enough energy to drive them at full power. In general, I think the checks could be re-organized into a set of warnings and errors:

* Errors:
  * **(no energy!)** No energy source (batteries or generator)
  * **(no steering!)** No steering.
  * **(no thruster!)** No thruster (ordinary thruster or afterburner).
  * **(overheating!)** Overheating at idle.
* Warnings:
  * **(afterburner only?)** Only mode of thrust is an afterburner.
  * **(battery only?)** Only energy source is a battery (will be stranded if battery wears out).
  * **(limited movement?)** No battery, and energy production insufficient for full thrusting or steering.
  * **(solar power?)** No battery, solar power only, and solar power far from the star will be below **{some criterion}**.

In the shops, errors and warnings could be shown with just a single icon above the ship icon:

![warning icons](https://user-images.githubusercontent.com/11450262/30751070-41cca59c-9f86-11e7-8371-f62f7fa4a28b.jpg)

Instead of adding a whole new class for ship warnings, a simpler approach would be to only report the most urgent error or warning for each ship. In that case, you wouldn't need a bitmask - you could just have the FlightCheck() function return a string, and the conversation names could be changed to the same strings. It would probably make sense for FlightCheck() to be a member of Ship.

To make things even simpler, the string could end in **!** or **?** to indicate whether it should use the error icon or the warning icon.

I'm not at all sure how XCode picks those GUID-like identifiers for new files you add, so I'm not sure if your manual edits to the project file to add the new classes will work or not.

Here are normal and high-dpi error icons:

![error](https://user-images.githubusercontent.com/11450262/30751777-4f6eca02-9f88-11e7-9228-bbb7c3f51a03.png)
![error 2x](https://user-images.githubusercontent.com/11450262/30751779-4f7062cc-9f88-11e7-96b0-d5fb1202ed95.png)
![warning](https://user-images.githubusercontent.com/11450262/30751778-4f700b24-9f88-11e7-811d-225cd6d52935.png)
![warning 2x](https://user-images.githubusercontent.com/11450262/30751780-4f7081a8-9f88-11e7-989a-450cc3ee4bd6.png)

flaviojs commented:
Since there is a distinction between errors and warnings, I presume the errors should prevent the ship from taking off, but not the warnings?

Only knowing one issue is a problem for me, it was one of the frustrations I had with the original conversation. The order of errors/warnings might be good for some builds but it will be bad for others (ex: my build giving priority to thrusting over steering VS the other way around). Since I can only safely store outfits by leaving them installed in parked ships, I end up having to shuffle around stuff from multiple ships when I want to change one of my ships.

I've been using the labels. By knowing the problematic ships and all the problems it has I can save a lot of time. I'll try putting all the issues in the tooltip again.

---

How is high-dpi handled by the interface? I can't find a example.

flaviojs commented:
Implemented the reorganized errors and warnings and used the icons.

Showing the icons separately in the ship details allows me to know immediately if I fixed a problem or created a new one. To know the warning details I need to hover over the icon, which is not a problem.

This looks good so next I'm gonna work on the conversations.

Ship list in the shop (hovering multi-warning icon and dragging ship):
![endless_sky_20170924_1](https://user-images.githubusercontent.com/1009600/30784876-35f7005c-a155-11e7-8277-a1a7c2478d48.png) ![endless_sky_20170924_2](https://user-images.githubusercontent.com/1009600/30784878-3908719a-a155-11e7-9391-d664dbb70912.png)

Ship details in the shop (hovering warning icon):
![endless_sky_20170924_3](https://user-images.githubusercontent.com/1009600/30784900-66e98892-a155-11e7-8be4-cb2a68e19bc4.png)

ShipInfoPanel (hovering warning icon):
![endless_sky_20170924_4](https://user-images.githubusercontent.com/1009600/30784904-7dcc3cda-a155-11e7-94fb-904aad1fdb50.png)

PlayerInfoPanel ("ship" column is shifted to the right):
![endless_sky_20170924_5](https://user-images.githubusercontent.com/1009600/30784906-8c078854-a155-11e7-9f79-34ad370819d3.png)

Bladewood commented on data/conversations.txt:
```diff
@@ -19,37 +19,175 @@ conversation "intro"
+	
+	label unknown
+	scene "outfit/unknown"
+	`You've heard enough horror stories about the sort of accidents that happen when starship systems malfunction to be completely convinced of the necessity of a pre-flight safety check. As you listen to the complex symphony of hums, rattles, and clicks of "<ship>" warming up, you run through a mental checklist. Computers, life support, engine...`
```
`Computers, life support, engine...`

engines*?

flaviojs replied:
Fixed.

flaviojs commented:
Moved flight checks from the outfitter to take off.

This is as far as I can take the conversation with the current capabilities.
Player conditions are used as input variables.
Each ship warning shows a conversation and it allow 4 actions:

- go to the outfitter: available if you can access the outfitter and the ship is in the system
- abort the take off: always available
- park the ship: available if it's not be the flagship
- dismiss the warning: available if it's not a serious warning

flaviojs commented:
@endless-sky 
I don't know what to do with the high dpi images.
All the stuff I mentioned in #2944 has been addressed, so it's no longer WIP.
Want me to rework the commits to remove discarded solutions and make the PR easier to evaluate?

Amazinite commented:
>I don't know what to do with the high dpi images.

Here: https://github.com/endless-sky/endless-sky-high-dpi

endless-sky commented:
I decided to just re-implement this in a simpler manner (abf8b78). Flight checks are a minor enough part of the game that I just don't think it makes sense to dedicate nearly a thousand lines of code to such a detailed presentation of them.
